sled-hardware/src/disk.rs

```diff
@@ -415,6 +546,53 @@ impl From<ZpoolKind> for DiskVariant {
     }
 }
 
+/// A file that wraps a zfs encryption key.
+///
+/// We put this in a RAM backed filesystem and zero and delete it when we are
+/// done with it. Unfortunately we cannot do this inside `Drop` because there is no
+/// equivalent async drop.
+pub struct KeyFile {
+    path: Keypath,
+    file: File,
+    log: Logger,
+}
+
+impl KeyFile {
+    pub async fn create(
+        path: Keypath,
+        key: &[u8; 32],
+        log: &Logger,
+    ) -> std::io::Result<KeyFile> {
+        // TODO: fix this to not truncate
+        // We want to overwrite any existing contents.
+        // If we truncate we may leave dirty pages around
+        // containing secrets.
+        let mut file = tokio::fs::OpenOptions::new()
+            .create(true)
+            .write(true)
+            .open(&path.0)
+            .await?;
+        file.write_all(key).await?;
+        info!(log, "Created keyfile {}", path);
+        Ok(KeyFile { path, file, log: log.clone() })
+    }
+
+    // It'd be nice to `impl Drop for `KeyFile` and then call `zero`
+    // from within the drop handler, but async `Drop` isn't supported.
```

Comment:
I don't actually think this does exactly what you want on most systems.

- If the filesystem is copy-on-write, this would just write new pages out, rather than actually removing old ones
- When we have swap, I think this file could be swapped back out to disk, unless `MAP_NORESERVE` was explicitly set on the region of memory used to hold the file

All this is to say: What do you think is the effect of writing zeroes out to the file before unlinking, as opposed to just unlinking the keyfile?

Reply:
The suggestions I gave are specific to the use of a tmpfs and the fact that `/var/run` is a tmpfs (which means `/var/run/oxide` is as well unless something really weird is going on).

* There is no copy-on-write going on in tmpfs. We're just copying data into pages and overwriting them.
* It is strictly true that this could get paged out against swap. `MAP_NORESERVE` however doesn't prevent swap. It just says that make this without a reservation and if I can't then you are allowed to kill me. (It's also an mmap concept and not really applicable to a file in a file system). However, if it does, ultimately this is what encryption of the swap device is for. This problem isn't really specific to this, in my opinion. The same series of things could happen with sled agent or zfs(8)'s memory. To prevent swapping you need to use mlock or similar. However, while that prevents paging, it again is specific to a processes's memory not a file that we're trying to read from.
* The zeroing concept specifically on tmpfs means that when this page is released on a tmpfs file system back to system memory pool, it won't be lingering somewhere with the data. Getting at that page would imply having access to physical memory, so it's maybe not that important, but just a kind of belt and suspenders kind of thing I would do as a moral equivalent to an `explicit_bzero()` or similar going on. This is definitely specific to the implementation of tmpfs. If the page was reused for another process it'd definitely be zeroed.

Reply:
I added some comments regarding this in [b9fef27](https://github.com/oxidecomputer/omicron/pull/3007/commits/b9fef273b22da301e87dbe5e1e353e7316d1ee04)